accepted/0000-sbom-command.md

```diff
@@ -0,0 +1,461 @@
+
+`--omit` - Dependency types to omit from generated SBOM. Valid values are “dev”, “optional”, and “peer” (can be set multiple times). By default, all development, optional, and peer dependencies will be included in the generated SBOM unless explicitly excluded.
+
+`--package-lock-only` - Constructs the SBOM based on the tree described by the _package-lock.json_, rather than the contents of _node_modules_. Defaults to _false_. If the _node_modules_ folder is not present, this flag will be required in order to generate an SBOM.
+ 
+`--workspace` - Runs the command only in the context of the specified workspace (can be set multiple times).
+
+`--workspaces` - Runs the command in the context of all configured workspaces. 
+
+If the user runs the `sbom` command without first installing the dependencies for the project (i.e. there is no _node_modules_ folder present) an error will be displayed. An SBOM can be generated solely based on the contents of the _package-lock.json_ but requires the user to explicitly specify the `--package-lock-only` flag.
+
+Initially, we'll support the most widely used versions of the SPDX and CycloneDX specifications (likely v2.3 for SPDX and v1.4 for CycloneDX). Best effort will be made to support new versions as they gain adoption across the ecosystem.
+
+
+## Rationale and Alternatives
+
+There are a few existing tools which can be used to generate an SBOM from an npm project:
+
+* <code>[@cyclonedex/cyclonedx-npm](https://www.npmjs.com/package/@cyclonedx/cyclonedx-npm)</code> - A CLI for generating a CycloneDX-style SBOM from an npm project. This project is written in TypeScript and actually invokes <code>npm-ls</code> in order to get dependency information for the project.
+* <code>[spdx-sbom-generator](https://github.com/opensbom-generator/spdx-sbom-generator)</code> - A CLI for generating SPDX-style SBOMs for a number of different package managers (including npm). Currently, this tool only works with npm projects using <em>lockfileVersion</em> 1 so it’s not viable for a large number of projects (current <em>lockfileVersion</em> is 3)
+
+While you can effectively generate the same output we’re proposing with this combination of tools, there is value in having this capability supported directly in npm. Beyond the obvious developer-experience benefit of having SBOM generation baked-in to the CLI, it gives us a future path to do things like automatic-signing of SBOMs or integration of SBOMs into the package publishing process.
+
+
+## Implementation
+
+The `npm-sbom` command is similar in function to `npm-ls` command and will likely utilize a similar implementation. We’ll use <code>[arborist](https://github.com/npm/cli/tree/latest/workspaces/arborist)</code> to construct the dependency tree and the <code>[treeverse](https://github.com/isaacs/treeverse)</code> library to traverse the tree and assemble the SBOM.
+
```

Comment:
you might use [@cyclonedx/cyclonedx-library](https://www.npmjs.com/package/@cyclonedx/cyclonedx-library) to generate the CycloneDX SBOM. written in Typescript, already used by mentioned [@cyclonedex/cyclonedx-npm](https://www.npmjs.com/package/@cyclonedx/cyclonedx-npm)

Reply:
I had considered proposing this, but I know that the npm team is quite intentional about the 3rd-party dependencies that get bundled with the CLI (carefully considering things like package size and long-term support). Given the deep dependency tree that `@cyclonedx/cyclonedx-library` brings with it, it seems unlikely that it would get accepted as a CLI dependency.

Reply:
I would just mention it as an argument how to simplify the implementation. Knowledge that there is a lib that does most things for you is a huge benefit when making decisions whether to accept a feature, don't you think?

It shall be to the developers to find reasons not to use an existing library. 

> Given the deep dependency tree that `@cyclonedx/cyclonedx-library` brings with it [...]
You are wrong. Only one new dependency. The `spdx-*` dependencies are already bundled with NPM.  The optional dependencies are not needed for the features described in your RFC. see 
```shellSession
$ npm i -S --omit optional @cyclonedx/cyclonedx-library
added 5 packages, and audited 6 packages in 10s

1 package is looking for funding
  run `npm fund` for details

found 0 vulnerabilities


$ npm ls --all
foo@1.0.0 /tmp/foo
└─┬ @cyclonedx/cyclonedx-library@4.0.0
  ├── UNMET OPTIONAL DEPENDENCY ajv-formats-draft2019@^1.6.1
  ├── UNMET OPTIONAL DEPENDENCY ajv-formats@^2.1.1
  ├── UNMET OPTIONAL DEPENDENCY ajv@^8.12.0
  ├── UNMET OPTIONAL DEPENDENCY libxmljs2@^0.31.0 || ^0.32.0
  ├── packageurl-js@1.0.2
  ├─┬ spdx-expression-parse@3.0.1
  │ ├── spdx-exceptions@2.3.0
  │ └── spdx-license-ids@3.0.13
  └── UNMET OPTIONAL DEPENDENCY xmlbuilder2@^3.0.2
```

Reply:
That's great info! I'll add a note to the RFC and the npm team can decide.